src/libraries/System.Text.Json/tests/Serialization/DictionaryTests.KeyConverter.cs

```diff
@@ -0,0 +1,431 @@
+        public async Task ThrowOnInvalidFormatAsync(string json, Type typeToConvert)
+        {
+            byte[] jsonBytes = Encoding.UTF8.GetBytes(json);
+            Stream stream = new MemoryStream(jsonBytes);
+
+            JsonException ex = await Assert.ThrowsAsync<JsonException>(async () => await JsonSerializer.DeserializeAsync(stream, typeToConvert));
+            Assert.Contains(typeToConvert.ToString(), ex.Message);
+        }
+
+        [Theory]
+        [InlineData(@"{""\u0039"":1}", typeof(Dictionary<int, int>), 9)]
+        [InlineData(@"{""\u0041"":1}", typeof(Dictionary<string, int>), "A")]
+        [InlineData(@"{""\u0066\u006f\u006f"":1}", typeof(Dictionary<MyEnum, int>), MyEnum.Foo)]
+        public static async Task TestUnescapedKeysAsync(string json, Type typeToConvert, object keyAsType)
+        {
+            byte[] utf8Json = Encoding.UTF8.GetBytes(json);
+            MemoryStream stream = new MemoryStream(utf8Json);
+
+            object result = await JsonSerializer.DeserializeAsync(stream, typeToConvert);
+
+            IDictionary dictionary = (IDictionary)result;
+            Assert.True(dictionary.Contains(keyAsType));
+        }
+
+        [Fact]
+        public static async Task TestUnescapedKeysAsync()
+        {
+            // Test Guid which cannot be passed as parameter on above method.
+            byte[] utf8Json = Encoding.UTF8.GetBytes(@"{""\u0036bb67e4e-9780-4895-851b-75f72ac34c5a"":1}");
+            MemoryStream stream = new MemoryStream(utf8Json);
+
+            Dictionary<Guid, int> result = await JsonSerializer.DeserializeAsync<Dictionary<Guid, int>>(stream);
+
+            Guid myGuid = new Guid("6bb67e4e-9780-4895-851b-75f72ac34c5a");
+            Assert.Equal(1, result[myGuid]);
+
+            // Test object.
+            utf8Json = Encoding.UTF8.GetBytes(@"{""\u0038"":1}");
+            stream = new MemoryStream(utf8Json);
+
+            Dictionary<object, int> result2 = await JsonSerializer.DeserializeAsync<Dictionary<object, int>>(stream);
+
+            Dictionary<object, int>.Enumerator enumerator = result2.GetEnumerator();
+            enumerator.MoveNext();
+
+            Assert.Equal(1, enumerator.Current.Value);
+
+            JsonElement dictionaryKey = (JsonElement)enumerator.Current.Key;
+            JsonElement myElement = JsonDocument.Parse("\"8\"").RootElement;
+
+            Assert.Equal(dictionaryKey.ValueKind, myElement.ValueKind);
+            Assert.Equal(dictionaryKey.GetString(), myElement.GetString());
+        }
+
+        private class UnsupportedDictionaryWrapper
+        {
+            public Dictionary<int[], int> Dictionary { get; set; }
+        }
+
+        private class UnsupportedDictionaryWrapper_Wrapper
+        {
+            public UnsupportedDictionaryWrapper Wrapper { get; set; }
+        }
+
+        [Fact]
+        public static void TestNotSuportedExceptionIsThrown()
```

Comment:
With this PR, `NotSupportedException`s are less often thrown since I don't check for the dictionary being supported **until** we actually call the dictionary `KeyConverter`. Is this behavior acceptable?

 @steveharter Does your changes in https://github.com/dotnet/runtime/pull/32669 related to delayed initialization also affect this behavior? 

I can see that I need to fix this to avoid behavior  changes but I just wanted to point this out since maybe this can be an acceptable or reasonable change.

The following table shows how this behave on master vs the feature branch.
Throws `NotSupportedException`?

| Type                                 | JSON                           | Branch: master | Branch: feature |
|--------------------------------------|--------------------------------|----------------|-----------------|
| Dictionary<int[], int>               |                                | yes            | yes             |
| Dictionary<int[], int>               | {}                             | yes            | yes             |
| Dictionary<int[], int>               | null                           | yes            | no              |
|                                      |                                |                |                 |
| UnsupportedDictionaryWrapper         |                                | yes            | no              |
| UnsupportedDictionaryWrapper         | {}                             | yes            | no              |
| UnsupportedDictionaryWrapper         | null                           | yes            | no              |
| UnsupportedDictionaryWrapper         | {"Dictionary":{}}              | yes            | yes             |
| UnsupportedDictionaryWrapper         | {"Dictionary":null}            | yes            | no              |
|                                      |                                |                |                 |
| UnsupportedDictionaryWrapper_Wrapper |                                | no             | no              |
| UnsupportedDictionaryWrapper_Wrapper | {}                             | no             | no              |
| UnsupportedDictionaryWrapper_Wrapper | null                           | no             | no              |
| UnsupportedDictionaryWrapper_Wrapper | {"Wrapper":{}}                 | yes            | no              |
| UnsupportedDictionaryWrapper_Wrapper | {"Wrapper":null}               | no             | no              |
| UnsupportedDictionaryWrapper_Wrapper | {"Wrapper":{"Dictionary":{}}   | yes            | yes             |
| UnsupportedDictionaryWrapper_Wrapper | {"Wrapper":{"Dictionary":null} | yes            | no              |
|                                      |                                |                |                 |
| List<Dictionary<int[], int>>         |                                | no             | no              |
| List<Dictionary<int[], int>>         | null                           | no             | no              |
| List<Dictionary<int[], int>>         | []                             | yes            | no              |
| List<Dictionary<int[], int>>         | [""]                           | yes            | no              |
| List<Dictionary<int[], int>>         | [null]                         | yes            | no              |
| List<Dictionary<int[], int>>         | [{}]                           | yes            | yes             |

cc @layomia

Reply:
Delayed validation makes sense to me. We do that in other cases as well for perf (including validating collection elements, elements of elements, etc).

As long as we continue to have tests for unsupported cases we should be fine. We should also check the Exception messages for the proper substrings (might require https://github.com/dotnet/runtime/pull/32669)